job-runner/chain/traverser.go

```diff
@@ -146,18 +165,31 @@ func (t *traverser) Run() error {
 
 	var finalState byte
 	defer func() {
-		// Set final state of chain in repo. This will be very short-lived because
-		// next we'll finalize the request in the RM. Although short-lived, we set
-		// it in case there's problems finalizing with RM.
-		t.chain.SetState(finalState)
-		t.chainRepo.Set(t.chain)
-
-		// Mark the request as finished in the Request Manager.
-		if err := t.rmc.FinishRequest(t.chain.RequestId(), finalState); err != nil {
-			t.logger.Errorf("problem reporting status of the finished chain: %s", err)
-		} else {
-			t.chainRepo.Remove(t.chain.RequestId())
+		// Only tell RM the request was finished if job chain wasn't suspended or
+		// if it was suspended but completed successfully anyways.
+		if !t.suspended() || finalState == proto.STATE_COMPLETE {
+			// Set final state of chain in repo. This will be very short-lived
+			// because next we'll finalize the request in the RM. Although
+			// short-lived, we set it in case there's problems finalizing with RM.
+			t.chain.SetState(finalState)
+			t.chainRepo.Set(t.chain)
+
+			// Mark the request as finished in the Request Manager.
+			if err := t.rmc.FinishRequest(t.chain.RequestId(), finalState); err != nil {
+				t.logger.Errorf("problem reporting status of the finished chain: %s", err)
+			} else {
+				t.chainRepo.Remove(t.chain.RequestId())
```

Comment:
For this blocks, let's add a `@todo: retry FinishRequest`.  I think we should always remove completed chains from repo, but in case RM is down or being redeployed and it temporarily fails, we should retry a few times. Right now, if it fails and we just log and leave chain in repo, I think that leaves phantom chains that are neither complete according to RM nor gone from chain repo.